File: ios/Runner/Alarm/AlarmEventBus.swift
import Flutter
import Foundation

final class AlarmEventBus: NSObject, FlutterStreamHandler {
    static let shared = AlarmEventBus()
    private var sink: FlutterEventSink?
    private var pending: [[String: Any]] = []

    private override init() {}

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        let snapshot = pending
        pending.removeAll()
        snapshot.forEach { events($0) }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }

    func emit(_ event: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if let sink = self.sink {
                sink(event)
            } else {
                self.pending.append(event)
            }
        }
    }
}

File: android/app/src/main/kotlin/com/example/count_to_three/alarm/AlarmEventBus.kt
package com.example.count_to_three.alarm

import android.os.Handler
import android.os.Looper
import com.example.count_to_three.alarm.model.AlarmEvent
import io.flutter.plugin.common.EventChannel

object AlarmEventBus : EventChannel.StreamHandler {
    private val lock = Any()
    private var sink: EventChannel.EventSink? = null
    private val pending = mutableListOf<AlarmEvent>()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        val snapshot: List<AlarmEvent>
        synchronized(lock) {
            sink = events
            snapshot = pending.toList()
            pending.clear()
        }
        snapshot.forEach { mainHandler.post { events?.success(it.toMap()) } }
    }

    override fun onCancel(arguments: Any?) {
        synchronized(lock) { sink = null }
    }

    fun emit(event: AlarmEvent) {
        synchronized(lock) {
            val s = sink
            if (s != null) {
                mainHandler.post { s.success(event.toMap()) }
            } else {
                pending.add(event)
            }
        }
    }
}
